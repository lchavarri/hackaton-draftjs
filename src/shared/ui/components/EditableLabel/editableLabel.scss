@import '../../assets/variables';

$editable-label-min-width: rem(40px);
$editable-label-max-width: rem(400px);
$editable-label-height: rem(22px);

// MOVE TO HIS OWN FILE
.char-count {
  border-radius: rem(7.5px);
  background-color: #f1f4fa;
  font-size: rem(11px);
  height: rem(15px);
  line-height: rem(15px);
  padding: rem(0 4px);

  + input:not([type='checkbox']):not([type='radio']) {
    padding-right: rem(40px);
  }
}

.editable-label {
  position: relative;
  display: inline-block;
  font-size: inherit;
  height: $editable-label-height;
  line-height: $editable-label-height;

  .char-count {
    display: none;
    position: absolute;
    right: 0;
    bottom: rem(3px);
  }

  &-input {
    position: absolute;
    top: rem(2px);
    display: inline-block;
    opacity: 0;
    pointer-events: none;
    transition: opacity $transition-fast;
    font-weight: inherit;
    min-width: $editable-label-min-width;
    max-width: $editable-label-max-width;
  }

  &-text {
    display: inline-block;
    text-overflow: ellipsis;
    overflow: hidden;
    max-width: $editable-label-max-width;
  }

  &-icon {
    width: $editable-label-height;
    height: 100%;
    vertical-align: top;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    font-size: rem(13px);
    position: relative;
    cursor: pointer;
    .spinner {
      position: absolute;
      top: rem(3px);
      left: rem(3px);
      width: rem(16px);
      height: rem(16px);
      margin: 0;
    }
    &:hover {
      color: $hover;
    }
  }

  // EDIT-MODE
  &.edit-mode .char-count {
    display: inline-flex;
  }

  &.edit-mode &-input {
    pointer-events: initial;
    position: static;
    opacity: 1;
  }

  &.edit-mode &-text {
    display: none;
  }

  &.edit-mode &-icon.edit {
    visibility: hidden;
  }
}
