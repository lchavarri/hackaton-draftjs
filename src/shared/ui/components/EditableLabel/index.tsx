import React, { useRef, useState, useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPen, faExclamationCircle } from '@fortawesome/pro-light-svg-icons';
import { Popup } from 'semantic-ui-react';

import Spinner from '../Spinner';
import './editableLabel.scss';

export type EditableLabelValidation = {
  valid: boolean;
  message?: string;
  externalValidation?: boolean;
  currentValue?: string;
};

type Props = {
  originalValue: string;
  onChange: Function;
  onValidation: (validation: EditableLabelValidation) => string;
  replaceSpaces?: boolean;
  charCountEnabled?: boolean;
  loading?: boolean;
  inputProps: {
    maxLength?: number;
    pattern?: string;
    placeholder?: string;
  };
  tooltipOnTop: boolean;
  isEditing?: boolean;
};

export default function EditableLabel(props: Props) {
  const {
    originalValue,
    onChange,
    onValidation,
    replaceSpaces,
    charCountEnabled,
    inputProps,
    loading: propsLoading,
    tooltipOnTop,
    isEditing
  } = props;

  const inputRef = useRef<any>(null);
  const [editing, setEditing] = useState(isEditing || false);
  const [loading, setLoading] = useState<boolean | undefined>(false);
  const [error, setError] = useState('');
  const [focused, setFocused] = useState(false);
  const [valueLength, setValueLength] = useState(
    originalValue ? originalValue.length : 0
  );

  useEffect(() => {
    setLoading(propsLoading);
  }, [propsLoading]);

  useEffect(() => {
    if (editing) {
      selectInputValue();
    }
    setFocused(editing);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [editing]);

  useEffect(() => {
    if (focused && inputRef && inputRef.current) {
      inputRef.current.focus();
    }
  }, [focused]);

  const selectInputValue = () => {
    setCustomValidity({ valid: true });
    inputRef.current.setSelectionRange(0, inputRef.current.value.length);
  };

  const setCustomValidity = (validation: EditableLabelValidation) => {
    if (inputRef && inputRef.current) {
      inputRef.current.setCustomValidity(
        validation.valid ? '' : validation.message
      );
    }
    const validationMessage = onValidation({
      ...validation,
      currentValue: inputRef.current && inputRef.current.value
    });
    setError(validationMessage);
  };

  const enterEdit = (e: any) => {
    setEditing(true);
    e.preventDefault();
  };

  const exitEdit = () => {
    if (inputRef.current.value === originalValue) {
      setEditing(false);
      return;
    }

    if (!inputRef.current.value) {
      inputRef.current.value = originalValue;
      setValueLength(originalValue.length);
      setEditing(false);
      return;
    }

    // invalid pattern
    if (inputRef.current.validity.patternMismatch) {
      return;
    }

    setLoading(true);
    setFocused(false);
    onChange(inputRef.current.value)
      .then(() => {
        setLoading(false);
        setEditing(false);
      })
      .catch((err: any) => {
        const message = err && err.message;
        setCustomValidity({ valid: false, message, externalValidation: true });
        setLoading(false);
        setFocused(true);
      });
  };

  const handleKeyDown = (e: any) => {
    if (e.key === 'Enter') {
      return exitEdit();
    }
    if (e.key === 'Escape') {
      inputRef.current.value = originalValue;
      setValueLength(originalValue.length);
      setCustomValidity({ valid: true });
      return exitEdit();
    }
  };

  const handleInput = () => {
    setValueLength(inputRef.current.value.length);
    if (replaceSpaces) {
      inputRef.current.value = inputRef.current.value.replace(' ', '_');
    }
    if (inputRef.current.validity.patternMismatch) {
      setCustomValidity({
        valid: false,
        message: 'Pattern mismatch',
        externalValidation: false
      });
    } else {
      setCustomValidity({ valid: true });
    }
  };

  const getIcon = () => {
    if (loading) {
      return (
        <span className="editable-label-icon">
          <Spinner primary />
        </span>
      );
    }
    return (
      <span className="editable-label-icon edit" onClick={enterEdit}>
        <FontAwesomeIcon className={'icon'} icon={faPen} />
      </span>
    );
  };

  const charCount = charCountEnabled ? (
    <span className="char-count">
      {valueLength}/{inputProps.maxLength}
    </span>
  ) : null;

  const getInput = () => {
    const input = (
      <input
        type="text"
        title=""
        ref={inputRef}
        className="editable-label-input material"
        defaultValue={originalValue}
        onKeyDown={handleKeyDown}
        onBlur={() => {
          setFocused(false);
          exitEdit();
        }}
        onFocus={() => setFocused(true)}
        onClick={e => {
          e.preventDefault();
          e.stopPropagation();
        }}
        onMouseDown={e => {
          e.stopPropagation();
        }}
        onInput={handleInput}
        disabled={loading}
        {...inputProps}
      />
    );

    const showPopup = error && error.length && focused;

    return (
      <Popup
        className={showPopup ? 'danger' : 'hidden'}
        hideOnScroll={true}
        size="tiny"
        wide="very"
        position={tooltipOnTop ? 'top center' : 'right center'}
        trigger={input}
        open={true}
        popperDependencies={[error, focused]}
      >
        <FontAwesomeIcon
          color="inherit"
          icon={faExclamationCircle}
          aria-label="Danger!"
        />
        &nbsp;{error}
      </Popup>
    );
  };

  return (
    <div className={`editable-label ${editing ? 'edit-mode' : ''} `}>
      {charCount}
      {getInput()}
      <span className="editable-label-text">{originalValue}</span>
      {getIcon()}
    </div>
  );
}
