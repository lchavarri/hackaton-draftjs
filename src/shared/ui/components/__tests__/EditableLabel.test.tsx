import React from 'react';
import { shallow, mount, ReactWrapper } from 'enzyme';
import EditableLabel from '../EditableLabel';

describe('EditableLabel component', () => {
  let component: ReactWrapper;
  let props: any;

  const defaultProps = {};
  const getComponent = () => <EditableLabel {...props} />;
  const wrap = () => shallow(getComponent());
  const wrapDeep = () => mount(getComponent());

  beforeEach(() => {
    props = defaultProps;
    component = wrap();
  });

  it('should match component snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
