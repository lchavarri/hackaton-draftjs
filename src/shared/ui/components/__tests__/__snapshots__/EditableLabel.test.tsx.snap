// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableLabel component should match component snapshot 1`] = `
<div
  className="editable-label  "
>
  <Popup
    className="hidden"
    disabled={false}
    hideOnScroll={true}
    offset={0}
    on="hover"
    open={true}
    pinned={false}
    popperDependencies={
      Array [
        "",
        false,
      ]
    }
    position="right center"
    size="tiny"
    trigger={
      <input
        className="editable-label-input material"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        title=""
        type="text"
      />
    }
    wide="very"
  >
    <mockConstructor
      aria-label="Danger!"
      color="inherit"
    />
    Â 
  </Popup>
  <span
    className="editable-label-text"
  />
  <span
    className="editable-label-icon edit"
    onClick={[Function]}
  >
    <mockConstructor
      className="icon"
    />
  </span>
</div>
`;
